File: Java/LinkedList/03_Singly_Circular_LinkedList_Java/SinglyCL_015.java
/*
================================================================================
Program Name    : InsertAtPos() in Singly Circular Linked List
Program Number  : 015
Description     : This program demonstrates the InsertAtPos() operation in a
                  Singly Circular Linked List using Java.
                  The program supports insertion at the beginning, end, and
                  any valid position in the list while maintaining the circular
                  link and the node count.
Date            : July 2026
Language        : Java
Author          : Ishwari Vijaykumar Surve
================================================================================
Concepts Used:
1. Object-Oriented Programming (OOP)
2. Class and Object
3. Constructor
4. Self-Referential Class
5. Singly Circular Linked List
6. InsertFirst()
7. InsertLast()
8. InsertAtPos()
9. Display Traversal using do-while Loop
10. Node Counting
11. Position Validation

Key Learning Points:
- InsertAtPos() inserts a node at any valid position.
- Position is validated before insertion.
- InsertFirst() is used when position is 1.
- InsertLast() is used when position is iCount + 1.
- For middle insertion, traversal reaches the previous node.
- Circular structure is preserved automatically.

================================================================================
*/
// Program 15 : InsertAtPos() in Singly Circular Linked List
class node
{
    public int data;
    public node next;

    node(int no)
    {
        this.data = no;
        this.next = null;
    }
}

class SinglyCL
{
    public node first;
    public node last;
    public int iCount;

    public SinglyCL()
    {
        this.first = null;
        this.last = null;
        this.iCount = 0;
    }

    ///////////////////////////////////////////////////////////////

    public void Display()
    {
        node temp = first;

        if(first == null && last == null)
        {
            return;
        }

        do
        {
            System.out.print("| " + temp.data + " | -> ");
            temp = temp.next;
        }while(temp != last.next);

        System.out.println();
    }

    ///////////////////////////////////////////////////////////////

    public int Count()
    {
        return iCount;
    }

    ///////////////////////////////////////////////////////////////

    public void InsertFirst(int no)
    {
        node newn = new node(no);

        if(first == null && last == null)
        {
            first = newn;
            last = newn;
        }
        else
        {
            newn.next = first;
            first = newn;
        }

        last.next = first;

        iCount++;
    }

    ///////////////////////////////////////////////////////////////

    public void InsertLast(int no)
    {
        node newn = new node(no);

        if(first == null && last == null)
        {
            first = newn;
            last = newn;
        }
        else
        {
            last.next = newn;
            last = newn;
        }

        last.next = first;

        iCount++;
    }

    ///////////////////////////////////////////////////////////////

    public void InsertAtPos(int no, int iPos)
    {
        if((iPos < 1) || (iPos > iCount + 1))
        {
            return;
        }

        if(iPos == 1)
        {
            InsertFirst(no);
        }
        else if(iPos == iCount + 1)
        {
            InsertLast(no);
        }
        else
        {
            node newn = new node(no);
            node temp = first;

            for(int i = 1; i < iPos - 1; i++)
            {
                temp = temp.next;
            }

            newn.next = temp.next;
            temp.next = newn;

            iCount++;
        }
    }

    ///////////////////////////////////////////////////////////////

    public void DeleteFirst()
    {
        if(first == null && last == null)
        {
            return;
        }
        else if(first == last)
        {
            first = null;
            last = null;
        }
        else
        {
            first = first.next;
            last.next = first;
        }

        iCount--;
    }

    ///////////////////////////////////////////////////////////////

    public void DeleteLast()
    {
        if(first == null && last == null)
        {
            return;
        }
        else if(first == last)
        {
            first = null;
            last = null;
        }
        else
        {
            node temp = first;

            while(temp.next != last)
            {
                temp = temp.next;
            }

            last = temp;
            last.next = first;
        }

        iCount--;
    }

    ///////////////////////////////////////////////////////////////

    public void DeleteAtPos(int iPos)
    {
        if((iPos < 1) || (iPos > iCount))
        {
            return;
        }

        if(iPos == 1)
        {
            DeleteFirst();
        }
        else if(iPos == iCount)
        {
            DeleteLast();
        }
        else
        {
            // Program 16
        }
    }
}

///////////////////////////////////////////////////////////////////////////

class SinglyCL15
{
    public static void main(String A[])
    {
        SinglyCL sobj = new SinglyCL();
        int iRet = 0;

        sobj.InsertFirst(51);
        sobj.InsertFirst(21);
        sobj.InsertFirst(11);

        sobj.InsertLast(101);
        sobj.InsertLast(111);
        sobj.InsertLast(121);

        sobj.Display();
        iRet = sobj.Count();
        System.out.println("Number of nodes are : " + iRet);

        sobj.DeleteFirst();
        sobj.Display();
        iRet = sobj.Count();
        System.out.println("Number of nodes are : " + iRet);

        sobj.DeleteLast();
        sobj.Display();
        iRet = sobj.Count();
        System.out.println("Number of nodes are : " + iRet);

        sobj.InsertAtPos(105, 4);
        sobj.Display();
        iRet = sobj.Count();
        System.out.println("Number of nodes are : " + iRet);
    }
}
/*
================================================================================
Expected Output:
================================================================================

| 11 | -> | 21 | -> | 51 | -> | 101 | -> | 111 | -> | 121 | ->
Number of nodes are : 6

| 21 | -> | 51 | -> | 101 | -> | 111 | -> | 121 | ->
Number of nodes are : 5

| 21 | -> | 51 | -> | 101 | -> | 111 | ->
Number of nodes are : 4

| 21 | -> | 51 | -> | 101 | -> | 105 | -> | 111 | ->
Number of nodes are : 5

================================================================================
Memory Representation:
================================================================================

Before InsertAtPos(105, 4)

                 first
                   │
                   ▼
        +----+      +----+      +-----+      +-----+
        | 21 | ───► | 51 | ───► | 101 | ───► | 111 |
        +----+      +----+      +-----+      +-----+
           ▲                                       │
           └────────────── last.next ──────────────┘
                                           ▲
                                           │
                                          last

After InsertAtPos(105, 4)

                 first
                   │
                   ▼
        +----+      +----+      +-----+      +-----+      +-----+
        | 21 | ───► | 51 | ───► | 101 | ───► | 105 | ───► | 111 |
        +----+      +----+      +-----+      +-----+      +-----+
           ▲                                                  │
           └──────────────────── last.next ───────────────────┘
                                                     ▲
                                                     │
                                                    last

================================================================================
Explanation:
================================================================================

The constructor initializes first and last to null and sets iCount to 0.

InsertFirst() inserts a new node at the beginning of the circular linked
list and updates last.next to maintain the circular connection.

InsertLast() inserts a new node at the end of the list, updates last to
the newly inserted node, and reconnects last.next to first.

InsertAtPos() first validates the given position.

• If the position is less than 1 or greater than iCount + 1, insertion is
  rejected.
• If the position is 1, InsertFirst() is called.
• If the position is iCount + 1, InsertLast() is called.
• Otherwise, traversal reaches the node before the desired position and the
  new node is inserted between two existing nodes.

For InsertAtPos(105, 4):

Step 1: Traverse to position 3 (node 101).

Step 2:
newn.next = temp.next

Step 3:
temp.next = newn

The circular link remains valid because the insertion occurs in the middle
of the list.

The node counter (iCount) is incremented after a successful middle insertion.

================================================================================
Program Flow:
================================================================================

Create SinglyCL Object
          │
          ▼
Constructor Invoked
          │
          ▼
InsertFirst()
          │
          ▼
InsertLast()
          │
          ▼
Display Linked List
          │
          ▼
DeleteFirst()
          │
          ▼
DeleteLast()
          │
          ▼
InsertAtPos(105, 4)
          │
          ▼
Display Updated List
          │
          ▼
Count Nodes
          │
          ▼
Program Terminates

================================================================================
Operations Performed:
================================================================================

1. Created a Singly Circular Linked List object.
2. Inserted nodes using InsertFirst().
3. Inserted nodes using InsertLast().
4. Displayed the circular linked list.
5. Counted the total number of nodes.
6. Deleted the first node.
7. Deleted the last node.
8. Inserted node 105 at position 4.
9. Displayed the updated linked list.
10. Updated iCount after insertion.

================================================================================
Time Complexity:
================================================================================

InsertFirst()   : O(1)
InsertLast()    : O(1)
InsertAtPos()   : O(n)
DeleteFirst()   : O(1)
DeleteLast()    : O(n)
Display()       : O(n)
Count()         : O(1)

Space Complexity: O(1)

================================================================================
Conclusion:
================================================================================

Program 015 demonstrates insertion at a specified position in a Singly
Circular Linked List. The operation correctly handles insertion at the
beginning, end, and middle of the list while preserving the circular
structure and maintaining an accurate node count.

================================================================================
*/

File: C++/LinkedList/03_Singly_Circular_Linear_LinkedList_CPP/SinglyCL_015.cpp
/*
===============================================================================
Program Name    : Singly Circular Linked List - InsertAtPos() Operation
Program Number  : 015
Description     : This program implements the InsertAtPos() function in a
                  Singly Circular Linked List. It allows insertion of a new
                  node at any valid position while maintaining the circular
                  link between the first and last nodes.

Date            : July 2026
Language        : C++
Author          : Ishwari Vijaykumar Surve
===============================================================================
Concepts Used:
1. Class and Objects
2. Self-Referential Structure
3. Dynamic Memory Allocation (new)
4. Pointer Traversal
5. Position Validation
6. InsertFirst()
7. InsertLast()
8. Pointer Manipulation
9. Circular Linked List

Key Learning Points:
- Validate the insertion position before performing insertion.
- Reuse InsertFirst() and InsertLast() whenever possible.
- Traverse the list until the previous node of the desired position.
- Update links without breaking the circular structure.
- Increase node count after successful insertion.

===============================================================================
*/
#include<iostream>
using namespace std;

#pragma pack(1)

struct node
{
    int data;
    struct node *next;
};

typedef struct node NODE;
typedef struct node *PNODE;

class SinglyCL
{
private:
    PNODE first;
    PNODE last;
    int iCount;

public:
    SinglyCL();

    void Display();
    int Count();

    void InsertFirst(int);
    void InsertLast(int);

    void DeleteFirst();
};

///////////////////////////////////////////////////////////////////////////

SinglyCL::SinglyCL()
{
    first = NULL;
    last = NULL;
    iCount = 0;
}

///////////////////////////////////////////////////////////////////////////

void SinglyCL::Display()
{
    if(first == NULL && last == NULL)
    {
        return;
    }

    PNODE temp = first;

    do
    {
        cout<<"| "<<temp->data<<" | -> ";
        temp = temp->next;
    }while(temp != last->next);

    cout<<endl;
}

///////////////////////////////////////////////////////////////////////////

int SinglyCL::Count()
{
    return iCount;
}

///////////////////////////////////////////////////////////////////////////

void SinglyCL::InsertFirst(int iNo)
{
    PNODE newn = new NODE;

    newn->data = iNo;
    newn->next = NULL;

    if(first == NULL && last == NULL)
    {
        first = newn;
        last = newn;
    }
    else
    {
        newn->next = first;
        first = newn;
    }

    last->next = first;

    iCount++;
}

///////////////////////////////////////////////////////////////////////////

void SinglyCL::InsertLast(int iNo)
{
    PNODE newn = new NODE;

    newn->data = iNo;
    newn->next = NULL;

    if(first == NULL && last == NULL)
    {
        first = newn;
        last = newn;
    }
    else
    {
        last->next = newn;
        last = newn;
    }

    last->next = first;

    iCount++;
}

///////////////////////////////////////////////////////////////////////////

void SinglyCL::DeleteFirst()
{
    if(first == NULL && last == NULL)
    {
        return;
    }
    else if(first == last)
    {
        delete first;
        first = NULL;
        last = NULL;
    }
    else
    {
        first = first->next;
        delete last->next;
        last->next = first;
    }

    iCount--;
}

void SinglyCL::DeleteLast()
{
    if(first == NULL && last == NULL)
    {
        return;
    }
    else if(first == last)
    {
        delete first;
        first = NULL;
        last = NULL;
    }
    else
    {
        PNODE temp = first;

        while(temp->next != last)
        {
            temp = temp->next;
        }

        delete last;
        last = temp;
        last->next = first;
    }

    iCount--;
}
void SinglyCL::InsertAtPos(int iNo, int iPos)
{
    if((iPos < 1) || (iPos > iCount + 1))
    {
        cout<<"Invalid Position\n";
        return;
    }

    if(iPos == 1)
    {
        InsertFirst(iNo);
    }
    else if(iPos == iCount + 1)
    {
        InsertLast(iNo);
    }
    else
    {
        PNODE newn = new NODE;
        PNODE temp = first;

        newn->data = iNo;
        newn->next = NULL;

        for(int i = 1; i < iPos - 1; i++)
        {
            temp = temp->next;
        }

        newn->next = temp->next;
        temp->next = newn;

        iCount++;
    }
}

///////////////////////////////////////////////////////////////////////////

int main()
{
    SinglyCL sobj;

    sobj.InsertFirst(51);
    sobj.InsertFirst(21);
    sobj.InsertFirst(11);

    sobj.InsertLast(101);
    sobj.InsertLast(111);
    sobj.InsertLast(121);

    cout<<"Linked List before deletion:"<<endl;
    sobj.Display();

    cout<<"Count : "<<sobj.Count()<<endl;

    sobj.DeleteFirst();

    cout<<"\nLinked List after DeleteFirst():"<<endl;
    sobj.Display();

    cout<<"Count : "<<sobj.Count()<<endl;

    sobj.DeleteLast();

    sobj.Display();
    iRet = sobj.Count();
    cout<<"Number of nodes are : "<<iRet<<"\n";

   sobj.InsertAtPos(105,4);

    sobj.Display();
    iRet = sobj.Count();
    cout<<"Number of nodes are : "<<iRet<<"\n";

    return 0;
}
/*
===============================================================================
Algorithm:
===============================================================================

1. Check whether the given position is valid.
2. If position is 1, call InsertFirst().
3. If position is iCount + 1, call InsertLast().
4. Otherwise:
      • Create a new node.
      • Traverse to (Position - 1).
      • Connect the new node between two existing nodes.
      • Increment node count.

===============================================================================
Visualization:
===============================================================================

Before Insertion (Insert 105 at Position 4)

First
  │
  ▼
+----+    +----+    +----+    +-----+    +-----+
| 11 | -> | 21 | -> | 51 | -> | 101 | -> | 111 |
+----+    +----+    +----+    +-----+    +-----+
   ^                                            │
   └────────────────────────────────────────────┘

After Insertion

First
  │
  ▼
+----+    +----+    +----+    +-----+    +-----+    +-----+
| 11 | -> | 21 | -> | 51 | -> | 105 | -> | 101 | -> | 111 |
+----+    +----+    +----+    +-----+    +-----+    +-----+
   ^                                                     │
   └─────────────────────────────────────────────────────┘

===============================================================================
Program Flow:
===============================================================================

Receive Data & Position
          │
          ▼
Validate Position
          │
          ▼
Position = 1 ?
     │Yes            No
     ▼               │
InsertFirst()        ▼
                 Position = Last + 1 ?
                     │Yes          No
                     ▼             ▼
                InsertLast()   Traverse List
                                     │
                                     ▼
                           Insert New Node
                                     │
                                     ▼
                             Increment Count
                                     │
                                     ▼
                                 Return

===============================================================================
Time Complexity:
===============================================================================

Best Case   : O(1)
Worst Case  : O(n)

Space Complexity : O(1)

===============================================================================
Conclusion:
===============================================================================

Program 015 implements insertion at any valid position in a Singly Circular
Linked List. It efficiently handles insertion at the beginning, end, and middle
of the list while preserving the circular nature of the linked list and updating
the node count correctly.

===============================================================================
*/
